docs/Dialects/FIRRTL/RationaleFIRRTL.md

```diff
@@ -569,6 +569,12 @@ The expected lowering for strict connects is for the connect to be eliminated an
 
 The reason we provide this foreign type support is to allow for partial lowering of FIRRTL to HW and other dialects. Passes might lower a subset of types and operations to the target dialect and we need a mechanism to have the lowered values be passed around the FIRRTL module hierarchy untouched alongside the FIRRTL ops that are yet to be lowered.
 
+### Const Types
+
+FIRRTL hardware types can be specified as `const`, meaning they can only be assigned compile-time constant values or values of other `const` types.
+
+All members of a `const` aggregate type must be `const`. A non-`const` aggregate can contain members that are both `const` and non-`const`.
```

Comment:
I conflicted about this, implementation wise.

Reply:
Why require this?  Isn't it just a matter of subwhatever (const agg(innerty)) -> const innerty ?

Reply:
I figured it would be less error-prone when directly interacting with const aggregate types for the members to be guaranteed to be const. Otherwise it's something that any code that deals with aggregate types might need to take into account.

Reply:
Also note, this isn't exactly a "requirement" as much as a result of how the implementation works. When constructing an aggregate type, the constness is determined by whether all members are const. So it's kind of the inverse of what this says.

Reply:
I updated the documentation to better reflect this

Reply:
I think explicit const flag on aggregate would be more consistent.  e.g. sign and unsigned are explicit even if you don't have bits for them to apply to.  This also simplifies type conversion at places:  just change the aggregate, don't change every field.

Reply:
This does simplify conversions and I like that this approach handles the empty bundle case. But I feel like there are going to be a bunch of places in the code base where we're checking all member constness when creating an aggregate type to determine if it should be `const`. 

Maybe we can just add an overload to `BundleType::create` and `FVectorType::create` for when we want to be explicit? I suppose we can start explicit and add implicit down the line if it's too cumbersome.

Reply:
Updated to make const aggregates explicit